datumaro/components/project.py

```diff
@@ -2,486 +2,2383 @@
+
+            if not data_dir:
+                if self._project._is_cached(obj_hash):
+                    data_dir = self._project.cache_path(obj_hash)
+                    cached = True
+                elif self._project._can_retrieve_from_vcs_cache(obj_hash):
+                    data_dir = self._project._materialize_obj(obj_hash)
+                    cached = True
+
+                if not data_dir or not osp.isdir(data_dir):
+                    log.debug("Build: skipping loading stage '%s' from "
+                        "cache obj '%s', because it is not available",
+                        current_name, obj_hash)
+                    return None
+
+            if data_dir:
+                assert osp.isdir(data_dir), data_dir
+                log.debug("Build: loading stage '%s' from '%s'",
+                    current_name, data_dir)
+                return ProjectSourceDataset(
+                    data_dir, self._tree, target, readonly=cached)
+
+            return None
+
+
+        graph = pipeline._graph
+        if len(graph) == 0:
+            raise EmptyPipelineError()
+
+        head = pipeline.head
+        if not head:
+            raise MissingPipelineHeadError()
+
+        # traverse the graph and initialize nodes from sources to the head
+        to_visit = [head]
+        while to_visit:
+            current_name = to_visit.pop()
+            current = graph.nodes[current_name]
+
+            assert current.get('dataset') is None
+
+            obj_hash = current['config'].hash
+            if obj_hash:
+                dataset = _try_load_from_cache(current_name, current)
+                if dataset is not None:
+                    current['dataset'] = dataset
+                    continue
+
+            uninitialized_parents = []
+            initialized_parents = []
+            if graph.in_degree(current_name) == 0:
+                if current['config'].type == 'source':
+                    source = self._tree.sources[
+                        self._tree.build_targets.strip_target_name(current_name)]
+                    if not source.is_generated:
+                        # Source is missing in the cache and cannot
+                        # be retrieved. It is assumed that all the
+                        # sources were downloaded earlier
+                        raise MissingObjectError("Failed to load target '%s': "
+                            "obj '%s' was not found in cache" % \
+                            (current_name, obj_hash))
+                elif current['config'].type == 'project':
+                    pass # pipeline is empty
                 else:
-                    s_config = config.sources[source_name]
-                    if s_config and s_config.format:
-                        # NOTE: consider imported sources as our own dataset
-                        path = None
+                    raise WrongSourceNodeError(
+                        "Unexpected leaf node '%s' of type '%s'" %
```

Comment:
If the node has no edges, wouldn't that make it a root node, rather than a leaf node?

Reply:
I don't know how to call it better, because we have a reversed tree here with multiple roots and a single top.